### PR TITLE
Add lax field and optional field access spec conformance tests

### PR DESCRIPTION
**Pull request by SandaruJayawardana**
Related lang issue [#32853](https://github.com/ballerina-platform/ballerina-lang/issues/32853)

### CODE REVIEW

**Comment by KavinduZoysa** [comment]
@SandaruJayawardana, 
can you add some tests to cover this,
```
The compile-time requirements on the optional-field-access-expr depend on whether the type descriptor describing T is lax:

if it is lax, then the only compile-time requirement is that M is non-empty;
if it is not lax, then, in addition, T must be a subtype of the union of () and the mapping basic type, and the type descriptor for T must include field-name as an individual-field-descriptor (if the type descriptor is a union, then this requirement must be satisfied by at least one member of the union).
```

File: conformance/lang/expressions/field-access-expression/field_access_expr_for_lax_type.balt [hunk 1]
```diff
@@ -0,0 +1,1093 @@
...
+    io:println(x4.__.__); // @output
+    io:println(x4.\..__); // @output 1
+}
+
+Test-Case: output
+Description: Test field access expression by accessing existing fields when they belong to JSON objects. Here, mapping
+             constructor contains computed-name-field. Test case 1.
+Labels: array-type, computed-name-field, DecimalNumber, error-type, field-access-expr, if-else-stmt, is-expr, json,
+        list-constructor-expr, logical-negation, mapping-constructor-expr, nil-literal, string, value:toBalString
+
```
Comment (SasinduDilshara) [inline conformance/lang/expressions/field-access-expression/field_access_expr_for_lax_type.balt:113]: `function-defn`, `functon-call-expr` labels are missing.

File: conformance/lang/expressions/field-access-expression/field_access_expr_for_lax_type.balt [hunk 2]
```diff
@@ -0,0 +1,1093 @@
...
+function getName() returns string {
+    return "nil";
+}
+
+Test-Case: output
+Description: Test field access expression by accessing existing fields when they belong to JSON objects. Here, mapping
+             constructor contains computed-name-field. Test case 2.
+Fail-Issue: ballerina-platform/ballerina-lang#33293
+Labels: array-type, computed-name-field, DecimalNumber, field-access-expr, json, list-constructor-expr,
+        mapping-constructor-expr, nil-literal, string
```
Comment (SasinduDilshara) [inline conformance/lang/expressions/field-access-expression/field_access_expr_for_lax_type.balt:134]: `function-defn`, `functon-call-expr` labels are missing.

File: conformance/lang/expressions/field-access-expression/field_access_expr_for_lax_type.balt [hunk 3]
```diff
@@ -0,0 +1,1093 @@
...
+    io:println(x2.a6); // @output 128
+}
+
+Test-Case: output
+Description: Test field access expression by accessing user-defined types when they belong to JSON objects.
+Fail-Issue: ballerina-platform/ballerina-lang#35105
+Labels: decimal, DecimalFloatingPointNumber, DecimalNumber, field-access-expr, float, FloatingPointTypeSuffix, int,
+        int:MAX_VALUE, int:SIGNED16_MIN_VALUE, int:SIGNED32_MAX_VALUE, int:SIGNED8_MAX_VALUE, json,
+        mapping-constructor-expr, module-type-defn, nil-literal, string, union-type
+
```
Comment (SasinduDilshara) [inline conformance/lang/expressions/field-access-expression/field_access_expr_for_lax_type.balt:332]: `null` label is missing.

File: conformance/lang/expressions/field-access-expression/field_access_expr_for_lax_type.balt [hunk 4]
```diff
@@ -0,0 +1,1093 @@
...
+    io:println(t14.b); // @output {"a":null}
+    io:println(t14.b.a); // @output
+    io:println(t14.c); // @output
+}
+
+Test-Case: output
+Description: Test field access expression by accessing existing fields when they belong to map<T>, where T is lax.
+Labels: boolean, computed-name-field, DecimalFloatingPointNumber, DecimalNumber, field-access-expr, float, json,
+        map-type, mapping-constructor-expr, nil-literal, spread-field, string
+
```
Comment (SasinduDilshara) [inline conformance/lang/expressions/field-access-expression/field_access_expr_for_lax_type.balt:417]: `function-defn`, `functon-call-expr` labels are missing.

File: conformance/lang/expressions/field-access-expression/field_access_expr_for_lax_type.balt [hunk 5]
```diff
@@ -0,0 +1,1093 @@
...
+    io:println(x2.a6.a6); // @output {"a6":128}
+}
+
+Test-Case: output
+Description: Test field access expression by accessing user-defined subtypes when they belong to map<T>. Here, T is lax.
+Fail-Issue: ballerina-platform/ballerina-lang#35105, #35134
+Labels: decimal, DecimalFloatingPointNumber, DecimalNumber, field-access-expr, float, FloatingPointTypeSuffix, int,
+        int:MAX_VALUE, int:SIGNED16_MIN_VALUE, int:SIGNED32_MAX_VALUE, int:SIGNED8_MAX_VALUE, json, map-type,
+        mapping-constructor-expr, module-type-defn, nil-literal, string, union-type
+
```
Comment (SasinduDilshara) [inline conformance/lang/expressions/field-access-expression/field_access_expr_for_lax_type.balt:727]: `null` label is missing.

File: conformance/lang/expressions/field-access-expression/field_access_expr_for_lax_type.balt [hunk 6]
```diff
@@ -0,0 +1,1093 @@
...
+    io:println(t14.b.a); // @output
+    io:println(t14.c); // @output
+}
+
+Test-Case: error
+Description: Test the static type of field-access-expr for lax type.
+Labels: array-type, boolean, decimal, DecimalFloatingPointNumber, DecimalNumber, error-type, field-access-expr, float,
+        FloatingPointTypeSuffix, HexFloatingPointLiteral, json, list-constructor-expr, mapping-constructor-expr,
+        member-access-expression, nil-literal, readonly-type, string
+
```
Comment (SasinduDilshara) [inline conformance/lang/expressions/field-access-expression/field_access_expr_for_lax_type.balt:821]: `null` label is missing.

File: conformance/lang/expressions/field-access-expression/field_access_expr_for_lax_type.balt [hunk 7]
```diff
@@ -0,0 +1,1093 @@
...
+    json|R x1 = {x: error("a")};
+    json|error y1 = (<json> x1).a.b.c.d.e; // @panic type 'R' cannot be cast to 'json'
+}
+
+Test-Case: error
+Description: Test if the expression does not have basic type mapping, the result is a new error value.
+Fail-Issue: ballerina-platform/ballerina-lang#35133
+Labels: array-type, boolean, DecimalNumber, field-access-expr, float, json, list-constructor-expr, mapping-constructor,
+        readonly-type, string
+
```
Comment (SasinduDilshara) [inline conformance/lang/expressions/field-access-expression/field_access_expr_for_lax_type.balt:885]: `null` label is missing.

File: conformance/lang/expressions/field-access-expression/optional_field_access_expr_lax_type.balt [hunk 8]
```diff
@@ -0,0 +1,1003 @@
+Test-Case: output
+Description: Test field access expression by accessing existing fields when they belong to JSON objects. Here, the
+             field-names are string-literals.
+Labels: array-type, boolean, decimal, DecimalFloatingPointNumber, DecimalNumber, field-access-expr, float,
+        FloatingPointTypeSuffix, HexFloatingPointLiteral, HexIntLiteral, json, list-constructor-expr,
+        mapping-constructor-expr, nil-literal, readonly-type, string
+
```
Comment (SasinduDilshara) [inline conformance/lang/expressions/field-access-expression/optional_field_access_expr_lax_type.balt:7]: Don't we need t=to add `optional-field-access-expr` label?

File: conformance/lang/expressions/field-access-expression/optional_field_access_expr_lax_type.balt [hunk 9]
```diff
@@ -0,0 +1,1003 @@
...
+    io:println(x4?.\?.?.__); // @output 1
+}
+
+
+Test-Case: output
+Description: Test field access expression by accessing existing fields when they belong to JSON objects. Here, mapping
+             constructor contains computed-name-field.
+Labels: array-type, computed-name-field, DecimalNumber, field-access-expr, json, list-constructor-expr,
+        mapping-constructor-expr, nil-literal, string
+
```
Comment (SasinduDilshara) [inline conformance/lang/expressions/field-access-expression/optional_field_access_expr_lax_type.balt:114]: `function-defn`, `functon-call-expr` labels are missing.

File: conformance/lang/expressions/field-access-expression/optional_field_access_expr_lax_type.balt [hunk 10]
```diff
@@ -0,0 +1,1003 @@
...
+    io:println(x2?.a5); // @output 0
+    io:println(x2?.a6); // @output 128
+}
+
+Test-Case: output
+Description: Test field access expression by accessing user-defined types when they belong to JSON.
+Fail-Issue: ballerina-platform/ballerina-lang#35105, #35107, #35108
+Labels: decimal, DecimalFloatingPointNumber, DecimalNumber, field-access-expr, float, FloatingPointTypeSuffix, int,
+        int:MAX_VALUE, int:SIGNED16_MIN_VALUE, int:SIGNED32_MAX_VALUE, int:SIGNED8_MAX_VALUE, json,
+        mapping-constructor-expr, module-type-defn, nil-literal, string, union-type
```
Comment (SasinduDilshara) [inline conformance/lang/expressions/field-access-expression/optional_field_access_expr_lax_type.balt:315]: `null` label is missing.

File: conformance/lang/expressions/field-access-expression/optional_field_access_expr_lax_type.balt [hunk 11]
```diff
@@ -0,0 +1,1003 @@
...
+    T5 t12 = null;
+    T5 t13 = "";
+
+    T6 t14 = {"a": 149, "b": {"a": null}, "c": null};
+
+    T7 t15 = 21;
+    T7 t16 = 1.2;
+    T7 t17 = "JSON";
+    T7 t18 = null;
+    T7 t19 = {"a": 1}; // runtime error
```
Comment (SasinduDilshara) [inline conformance/lang/expressions/field-access-expression/optional_field_access_expr_lax_type.balt:351]: Can we remove this comment.

File: conformance/lang/expressions/field-access-expression/optional_field_access_expr_lax_type.balt [hunk 12]
```diff
@@ -0,0 +1,1003 @@
...
+    io:println(t14?.b); // @output {"a":null} // error NPE
+    io:println(t14?.b?.a); // @output // error JVM error
+    io:println(t14?.c); // @output
+}
+
+Test-Case: output
+Description: Test field access expression by accessing existing fields when they belong to map<T>, where T is lax.
+Labels: boolean, computed-name-field, DecimalFloatingPointNumber, DecimalNumber, field-access-expr, float, json,
+        map-type, mapping-constructor-expr, nil-literal, spread-field, string
+
```
Comment (SasinduDilshara) [inline conformance/lang/expressions/field-access-expression/optional_field_access_expr_lax_type.balt:401]: missing `int` label.

File: conformance/lang/expressions/field-access-expression/optional_field_access_expr_lax_type.balt [hunk 13]
```diff
@@ -0,0 +1,1003 @@
...
+    io:println(x2?.a5?.a5); // @output {"a5":0}
+    io:println(x2?.a6?.a6); // @output {"a6":128}
+}
+
+Test-Case: output
+Description: Test field access expression by accessing user-defined subtypes when they belong to map<T>. Here, T is lax.
+Fail-Issue: ballerina-platform/ballerina-lang#35105, #35107, #35108
+Labels: decimal, DecimalFloatingPointNumber, DecimalNumber, field-access-expr, float, FloatingPointTypeSuffix, int, 
+        int:MAX_VALUE, int:SIGNED16_MIN_VALUE, int:SIGNED32_MAX_VALUE, int:SIGNED8_MAX_VALUE, json, map-type, 
+        mapping-constructor-expr, module-type-defn, nil-literal, string, union-type
```
Comment (SasinduDilshara) [inline conformance/lang/expressions/field-access-expression/optional_field_access_expr_lax_type.balt:710]: `null` label is missing.

File: conformance/lang/expressions/field-access-expression/field_access_expr_for_lax_type.balt (hunk 1 above)
Comment (SandaruJayawardana) [reply conformance/lang/expressions/field-access-expression/field_access_expr_for_lax_type.balt:113]: Added `null`. Here function-call-expr is not tested explicitly and it belongs to minimal subset. Therefore, we can ignore these labels.

File: conformance/lang/expressions/field-access-expression/field_access_expr_for_lax_type.balt (hunk 2 above)
Comment (SandaruJayawardana) [reply conformance/lang/expressions/field-access-expression/field_access_expr_for_lax_type.balt:134]: https://github.com/ballerina-platform/ballerina-spec/pull/1048#discussion_r918548845

File: conformance/lang/expressions/field-access-expression/field_access_expr_for_lax_type.balt (hunk 4 above)
Comment (SandaruJayawardana) [reply conformance/lang/expressions/field-access-expression/field_access_expr_for_lax_type.balt:417]: https://github.com/ballerina-platform/ballerina-spec/pull/1048#discussion_r918548845

File: conformance/lang/expressions/field-access-expression/optional_field_access_expr_lax_type.balt (hunk 8 above)
Comment (SandaruJayawardana) [reply conformance/lang/expressions/field-access-expression/optional_field_access_expr_lax_type.balt:7]: My bad. Will update `field-access-expr` to `optional-field-access-expr` for these tests. Thanks.

File: conformance/lang/expressions/field-access-expression/optional_field_access_expr_lax_type.balt (hunk 9 above)
Comment (SandaruJayawardana) [reply conformance/lang/expressions/field-access-expression/optional_field_access_expr_lax_type.balt:235]: https://github.com/ballerina-platform/ballerina-spec/pull/1048#discussion_r918548845

File: conformance/lang/expressions/field-access-expression/optional_field_access_expr_lax_type.balt (hunk 12 above)
Comment (SandaruJayawardana) [reply conformance/lang/expressions/field-access-expression/optional_field_access_expr_lax_type.balt:522]: https://github.com/ballerina-platform/ballerina-spec/pull/1048#discussion_r918548845

**Comment by SandaruJayawardana** [comment]
> can you add some tests to cover this,
```
The compile-time requirements on the optional-field-access-expr depend on whether the type descriptor describing T is lax:
if it is lax, then the only compile-time requirement is that M is non-empty;
if it is not lax, then, in addition, T must be a subtype of the union of () and the mapping basic type, and the type descriptor for T must include field-name as an individual-field-descriptor (if the type descriptor is a union, then this requirement must be satisfied by at least one member of the union).
```

Added in https://github.com/ballerina-platform/ballerina-spec/pull/1048/commits/cf73865c825ffcd53e0334e6aec8c3323c9d3281 & https://github.com/ballerina-platform/ballerina-spec/pull/1048/commits/89f84e4a43cd5b5056a92fe3f16e2279d18d4816.

**Pull request closed without merging**